chain/events/filter/index.go

```diff
@@ -520,6 +609,13 @@ func (ei *EventIndex) Close() error {
 	return ei.db.Close()
 }
 
+func (ei *EventIndex) isTipsetProcessed(ctx context.Context, tipsetKey []byte) (bool, error) {
+	row := ei.db.QueryRowContext(ctx, "SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key = ?", tipsetKey)
```

Comment:
I think we need an additional "reverted" flag if we're going to use this table to fix https://github.com/filecoin-project/lotus/issues/11640. The issue here is:

1. Apply TipSet A at epoch X (is process returns false).
2. Switch to TipSet B at epoch X (is processed returns false).
4. Switch back to TipSetA (is processed returns _true_, but it's not).

When I discussed this with @rvagg the plan was (IIRC) to record:

1. The earliest processed tipset/epoch for backfilling.
2. The last processed tipset (and _only_ the last processed one).

When "catching up", you'd always look at the "path" between the last processed tipset and the current tipset (taking reverts into account).

But that doesn't solve your `getLogs` case. If a user asks for a historical "uncle", there's no easy way to tell if that uncle has been processed (you don't know what chain it's on).

So, I think the solution is to add a reverted flag so, on start, we can find the latest non-reverted tipset and "catch up" from there.

---

Note: We don't have to do this now, but I'd like to clean all this up a bit:

1. Remove the reverted flag from the `events` table, that'll be recorded in the `events_seen` table (maybe call this `tipsets`?).
2. Remove the `tipset_cid`/`tipset_key` from the `events` table as well and reference the correct entry in the `tipsets` table by foreign key.

It makes some of the queries more complex, but the events table is getting fat and we're doing more work than necessary to process reverts.

Reply:
I'm not sure we need a "reverted" column specifically for this use-case, we just need to know whether a particular tipset has been inserted or not. Even if it flips from applied to reverted and back to applied again, the tipset doesn't change so it's events won't change. What am I not seeing that #11640 would need a "reverted" @Stebalien? It makes sense if we have a "tipset" table where we store all the tipset things, but that's probably too much scope for this.

Reply:
You're right, it fixes the main issue but not my followup comment (https://github.com/filecoin-project/lotus/issues/11640#issuecomment-1964955238).

1. Main issue: Sometimes a user will call `getLogs` on a tipset we haven't processed yet. This PR fixes that case (or works towards fixing it).
2. Followup comment: Sometimes we'll shut down before indexing some tipset and, when we restart, we need to "catch up" and index any missing tipsets. In this case, my concern is that we'll shutdown right after step 3 above ("Switch back to TipSetA"), restart, and think "oh, I see we've already indexed TipSetA". But we _actually_ need to go and un-mark TipSetA as reverted and mark TipSetB as reverted instead.

Reply:
I think I agree with Steb.

Being able to answer "What's the latest tipset for which you have events on the cannonical chain" is important. Might even be for this work itself. Still working through it. But have made this change.

Reply:
This is done.